### ISSUE TITLE
ttnn.batch_norm low PCC in vanilla unet

### ISSUE THREAD

**Issue by keerthana-r-mcw** [tenstorrent/tt-metal#21363]
### Describe the bug

ttnn.batch_norm fails with assertion error in unit test and in the model.

### Steps to reproduce the issue

1. checkout [keerthanar/vanilla_unet_issues](https://github.com/tenstorrent/tt-metal/tree/keerthanar/vanilla_unet_issues)
2. Use the following command to run the unit test, ```pytest tests/ttnn/unit_tests/operations/fused/test_batch_norm.py::test_batch_norm_vanilla_unet```
3. Use the following command to run the model, ```pytest tests/ttnn/integration_tests/vanilla_unet/test_ttnn_unet.py```

### Please complete the following environment information

WH n150 or n300

**Comment by VirdhatchaniKN** [comment]
@keerthana-r-mcw ,Just went through the test files.
Will check the PCC Drop in the unit test, but I do not see the bad variant access in the model file. I guess its related to sharding. Can you please check and update?

**Comment by keerthana-r-mcw** [comment]
> [@keerthana-r-mcw](https://github.com/keerthana-r-mcw) ,Just went through the test files. Will check the PCC Drop in the unit test, but I do not see the bad variant access in the model file. I guess its related to sharding. Can you please check and update?

@VirdhatchaniKN  bad variant access is resolved but the model is giving low pcc (-0.003144934078669592)

**Comment by VirdhatchaniKN** [comment]
I'll look into this PCC Drop issue. 
Thanks @keerthana-r-mcw

**Comment by VirdhatchaniKN** [comment]
Hi @keerthana-r-mcw , I went through one of the files and below is my analysis. I expect the same issue in the second file as well. 

Debug analysis of the test file : `tests/ttnn/unit_tests/operations/fused/test_batch_norm.py::test_batch_norm_vanilla_unet` is as below

### Issues causing assertion error
- Mismatching modes of operation used for computing torch, TTNN Batch norm.
- Incorrect mean and variance computation in training mode : https://github.com/tenstorrent/tt-metal/issues/23876
- One other point to note is the data type mismatch. Torch is computed in FP32 , But TTNN is in BF16 (PCC Drop still persist on correcting this as-well, due to incorrect mean and variance results)

### Torch vs TTNN Input details passed in test file with results

- **Without changes**

<img width="837" alt="Image" src="https://github.com/user-attachments/assets/b6f583ec-88c8-492a-8cd5-93156d832e00" />

- **With changes - Matching dtypes(FP32) and mode of operation(Training mode)**

<img width="822" alt="Image" src="https://github.com/user-attachments/assets/f6126aeb-a084-495b-b5ee-aa1267703e39" />

### Detailed description of the above mentioned issues : 

- Default mode set in TTNN is evaluation(Inference) mode but torch runs in training mode 
  - Ref used for BN Implementation : [Link](https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.batch_norm.html), but [BatchNorm2d](https://docs.pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html#torch.nn.BatchNorm2d) requires .train() or .eval(). Since none was specified, the default is training mode. 
  - **Mismatching modes of operation :**
    - When in training mode, the stats value are computed from the input tensor (batch stat). They are then updated as running mean and running variance. 
    - This updated running stats value is what was passed to TTNN as inputs. But this updated stat value was not used in pytorch computation for BN, i.e., TTNN Requires to be passed the batch stats value but was passed with updated running stats value instead.
      - Even if this was corrected, since the default mode in `ttnn.batch_norm` is evaluation mode, the computation would have resulted in incorrect results. 
      - Hence, Mismatching modes of operation.
  - **Data type mismatch :** As in the above screenshot of the test file inputs, since the value passed was using torch.full, PCC Drop possibly would not happen because of this(Explained for both dtypes below). Also float32 results are precise compared to bfloat16.
  - **Incorrect mean and variance computation in training mode**
    - On ensuring both torch and TTNN BN works on same mode of operation, Assertion issue persist due to incorrect mean
    - For a tensor filled with 1.0, it computes 
      - mean = 1.0
      - variance = 0.0
      - Hence (input - batch_mean) = 1-1 = 0 --> which is expected Torch BN Result.
    - When set to training mode(In TTNN and torch), the result expected is 0, but TTNN Gives 
      - FP32 → 0.05411 
        - Computed Mean = 0.99707
        - Computed Var = 0.00292
      - BF16 → 0.06250 
        - Computed Mean = 0.99609
        - Computed Var = 0.00391
      - Considering FP32 Case mean and variance, the formula for BN computation is as follows:
        - (input - batch_mean)* [ 1/(sqrt(batch_var + eps))]  * weight + bias
        - (1.0 - 0.99707)* [ 1/(sqrt(0.0029211 + 1e-05))]  * 1 + 0 = (0.00293)* [ 1/(sqrt(0.0029311))] * 1 + 0 = 0.00293 * 18.4707 * 1 + 0 = 0.05411 * 1 + 0 = 0.05411 (No issues with computation, but with incorrect mean variance generation)
    - Thus, the drop is caused due to mean, variance in BN, which is computed using `ttnn.mean` (Refer [ttnn/cpp/ttnn/operations/normalization/batch_norm/batch_norm.cpp](https://github.com/tenstorrent/tt-metal/blob/050ef90563b6d99c6cc46530cf2748190ae0290c/ttnn/cpp/ttnn/operations/normalization/batch_norm/batch_norm.cpp))

**Comment by keerthana-r-mcw** [comment]
Hi @VirdhatchaniKN , This could be a possible reason for the unit test failure. However, the BatchNorm layer is in evaluation mode in the model test. Additionally, I observed a PCC of 0.99 for the tensor before this specific BatchNorm layer, and I verified that the mean and variance are identical between TTNN and Torch. I’m attaching the input and parameter dumps for reference.
https://drive.google.com/file/d/17jO2jVGdlh2lsVdJoAfx1GaeBUjtfKTm/view?usp=sharing

**Comment by saichandax** [comment]
@VirdhatchaniKN , please share an further update. This is a blocker for Unet Vanilla model.
cc: @mbahnasTT , @dvartaniansTT

**Comment by VirdhatchaniKN** [comment]
Update : 
Since the input tensor has a large shape, I wasn’t able to directly reuse it with BatchNorm. So, I tried a few alternative methods:

- Generated a tensor with the same range as the original input tensor (which was causing the PCC drop in the model) using the given stat values. However, I didn’t observe any drop in that setup.
- Tried printing the difference(Torch-Expected) and extracted the maximum and minimum difference. The idea was to isolate the corresponding input values that might be causing the discrepancy. This part is still in progress, but we’re currently facing an inconsistency:
  - Inside the model , the max and min difference is as follows : 
    - Min value: -0.42865490913391113 
    - Max value: 0.25357604026794434
      - This difference with bf16 data type is okay I guess.
  - Outside, the max and min is as follows : 
    - Min value: -3.816535711288452
    - Max value: 3.737806558609009
We expect these to match, but they don’t.

**Comment by keerthana-r-mcw** [comment]
Adding .eval() to that particular Batchnorm in the torch pipeline resolved the issue. Thanks
